packages/flutter/test/widgets/text_selection_test.dart

```diff
@@ -380,4 +405,213 @@ void main() {
   });
+
+  testWidgets('test EditableTextSelectionHandlersProvider long press', (WidgetTester tester) async {
+    await pumpEditableTextSelectionHandlersProvider(tester);
+    final TestGesture gesture =
+      await tester.startGesture(const Offset(200.0, 200.0), pointer: 0, kind: PointerDeviceKind.touch);
+    await tester.pump(const Duration(seconds: 2));
+    await gesture.up();
+    await tester.pumpAndSettle();
+
+    final FakeEditableTextState state = tester.state(find.byType(FakeEditableText));
+    final FakeRenderEditable renderEditable = tester.renderObject(find.byType(FakeEditable));
+    expect(state.showToolbarCalled, isTrue);
+    expect(renderEditable.selectPositionAtCalled, isTrue);
+  });
+
+  testWidgets('test EditableTextSelectionHandlersProvider tap', (WidgetTester tester) async {
+    await pumpEditableTextSelectionHandlersProvider(tester);
+    final TestGesture gesture =
+      await tester.startGesture(const Offset(200.0, 200.0), pointer: 0, kind: PointerDeviceKind.touch);
+    await gesture.up();
+    await tester.pumpAndSettle();
+
+    final FakeEditableTextState state = tester.state(find.byType(FakeEditableText));
+    final FakeRenderEditable renderEditable = tester.renderObject(find.byType(FakeEditable));
+    expect(state.showToolbarCalled, isFalse);
+    expect(renderEditable.selectWordEdgeCalled, isTrue);
+  });
+
+  testWidgets('test EditableTextSelectionHandlersProvider double tap', (WidgetTester tester) async {
+    await pumpEditableTextSelectionHandlersProvider(tester);
+    final TestGesture gesture =
+      await tester.startGesture(const Offset(200.0, 200.0), pointer: 0, kind: PointerDeviceKind.touch);
+    await tester.pump(const Duration(milliseconds: 50));
+    await gesture.up();
+    await gesture.down(const Offset(200.0, 200.0));
+    await tester.pump(const Duration(milliseconds: 50));
+    await gesture.up();
+    await tester.pumpAndSettle();
+
+    final FakeEditableTextState state = tester.state(find.byType(FakeEditableText));
+    final FakeRenderEditable renderEditable = tester.renderObject(find.byType(FakeEditable));
+    expect(state.showToolbarCalled, isTrue);
+    expect(renderEditable.selectWordCalled, isTrue);
+  });
+
+  testWidgets('test EditableTextSelectionHandlersProvider forcePress', (WidgetTester tester) async {
+    await pumpEditableTextSelectionHandlersProvider(tester);
+    final TestGesture gesture = await tester.createGesture();
+    await gesture.downWithCustomEvent(
+      const Offset(200.0, 200.0),
+      const PointerDownEvent(
+        pointer: 0,
+        position: Offset(200.0, 200.0),
+        pressure: 3.0,
+        pressureMax: 6.0,
+        pressureMin: 0.0,
+      ),
+    );
+    await gesture.up();
+    await tester.pump();
+
+    final FakeEditableTextState state = tester.state(find.byType(FakeEditableText));
+    final FakeRenderEditable renderEditable = tester.renderObject(find.byType(FakeEditable));
+    expect(state.showToolbarCalled, isFalse);
```

Comment:
This feels like a bug to me, the downWithCustomEvent does not trigger an onTapDown. It directly trigger forcepress event without _shouldShowSelectionToolbar been set. which end up in this state. @justinmc is this intended?

Reply:
in your example yes, only onForcePressStart will trigger.
for handling events differently, I am not sure if I still follow which events were you refer to?

Reply:
Does this seem good to you now, or do you still worry about a bug?

I was just saying that if the behavior of the onForcePressStart and onTapDown callbacks look correct, and you still think there is a bug, then maybe the bug lies in how we respond to those gestures in EditableTextSelectionHandlersProvider.

Reply:
I think the callbacks are correct, and I can live with the current behavior